src/transformers/models/realm/modeling_realm.py

```diff
@@ -0,0 +1,1757 @@
+            end_positions = end_positions.clamp(-1, ignored_index)
+
+            retriever_correct = has_answers
+            any_retriever_correct = torch.any(retriever_correct)
+
+            reader_correct = compute_correct_candidates(
+                candidate_starts=candidate_starts,
+                candidate_ends=candidate_ends,
+                gold_starts=start_positions[0 : self.config.reader_beam_size],
+                gold_ends=end_positions[0 : self.config.reader_beam_size],
+            )
+            any_reader_correct = torch.any(reader_correct)
+
+            retriever_loss = marginal_log_loss(relevance_score, retriever_correct)
+            reader_loss = marginal_log_loss(reader_logits.view(-1), reader_correct.view(-1))
+            retriever_loss *= any_retriever_correct.type(torch.float32)
+            reader_loss *= any_reader_correct.type(torch.float32)
+
+            total_loss = (retriever_loss + reader_loss).mean()
+
+        if not return_dict:
+            output = (predicted_block_index, predicted_candidate, predicted_start, predicted_end) + outputs[2:]
+            return (
+                ((total_loss, retriever_loss, reader_loss, retriever_correct, reader_correct) + output)
+                if total_loss is not None
+                else output
+            )
+
+        return RealmReaderOutput(
+            loss=total_loss,
+            retriever_loss=retriever_loss,
+            reader_loss=reader_loss,
+            retriever_correct=retriever_correct,
+            reader_correct=reader_correct,
+            block_idx=predicted_block_index,
+            candidate=predicted_candidate,
+            start_pos=predicted_start,
+            end_pos=predicted_end,
+            hidden_states=outputs.hidden_states,
+            attentions=outputs.attentions,
+        )
+
+
+REALM_FOR_OPEN_QA_DOCSTRING = r"""
+    Args:
+        question (`str`):
+            OpenQA Question.
+        answer_ids (`torch.LongTensor` of shape `(num_answers, answer_length)`, *optional*):
+            Answer ids for computing the marginal log-likelihood loss. Indices should be in `[-1, 0, ..., config.vocab_size]` (see `input_ids` docstring) Tokens with indices set to `-1` are ignored (masked),
+            the loss is only computed for the tokens with labels in `[0, ..., config.vocab_size]`
+        return_dict (`bool`, *optional*):
+            Whether or not to return a [`~file_utils.ModelOutput`] instead of a plain tuple.
+"""
+
+
+@add_start_docstrings(
+    "`RealmForOpenQA` for end-to-end open domain question answering.",
+    REALM_START_DOCSTRING,
+)
+class RealmForOpenQA(RealmPreTrainedModel):
+    def __init__(self, config, retriever=None):
+        super().__init__(config)
+        self.embedder = RealmEmbedder(config)
+        self.reader = RealmReader(config)
+        self.block_emb = nn.Parameter(torch.FloatTensor(size=(config.num_block_records, config.retriever_proj_size)).uniform_())
```

Comment:
@patrickvonplaten - The `block_emb` was intended to not be a parameter because it should not participate in gradient calculations; otherwise, this would contribute to a huge memory cost especially when running on GPUs. I instead registered it as a buffer in order to save it with other parameters into a checkpoint.

```python
        self.register_buffer(
            "block_emb",
            torch.zeros(()).new_empty(
                size=(config.num_block_records, config.retriever_proj_size),
                dtype=torch.FloatTensor,
                device=torch.device("cpu"),
            ),
        )
```

Here is the original TF impl. which uses a local variable to store this tensor and doesn't save it along with other parameters.

https://github.com/google-research/language/blob/master/language/orqa/utils/scann_utils.py#L39-L64

Reply:
I see this makes sense! Do you think we could instead maybe just set:

`self.block_emb.requires_grad = False` to remove this param from the gradient computation?

Reply:
IMO saving it as a paramter is slightly better as it's a "fixed parameter" that cannot be set deterministically no? So I think the best solution here would be to simple se the `requires_grad` to False no?

Reply:
Sorry I'm not quite sure what you meant by `cannot be set deterministically`. I think the `block_emb` should always be set deterministically? otherwise the `ForOpenQA` model would become useless if the `block_emb` is initialized randomly as the retrieval process relies on this pre-computed tensor to retrieve answer contexts from the `block_records`.

Although the two approaches have no difference in terms of model training or inference, users will see this parameter which actually behave like a buffer when they're calling `model.parameters()`; also, the optimizer will have to unnecessarily skip this parameter in the `step` method. 

Here is a discussion related to the usage of parameter and buffer: https://discuss.pytorch.org/t/what-is-the-difference-between-register-buffer-and-register-parameter-of-nn-module/32723/11

What do you think?

Reply:
Ok happy to change it to a buffer if you feel strongly about it! 

Sorry my explanation was probably not very clear. We use buffers in `transformers` mostly for tensors that are never any trainable weights are parameters adapted to some data. *E.g.*: https://github.com/huggingface/transformers/blob/f2ab21833f3c1fcc8dea76988bdacd368fca779e/src/transformers/models/gpt2/modeling_gpt2.py#L140 is the causal mask for GPT2. However `block_emb` is adapted (or at least dependent) on the open-qa dataset here which to me is more of a parameter than a buffer. 

However, I really don't feel strongly about this and if you prefer to save the variable as a buffer then let's go for this :-) As stated in the link you provided it shouldn't make a big difference

Reply:
Thanks for the clarification, I totally get your point. Currently I just set `block_emb` as a buffer, if there is any flew of this setting occurring, I'm fine to turn it to a `Parameter` :-)